pkg/datalayer/datalayer.go

```diff
@@ -1,28 +1,172 @@
+}
+
+func NewDataLayer(basePath string) *DataLayer {
+	return &DataLayer{basePath}
+}
+
+func (dl *DataLayer) volumePath(volumeName string) string {
+	return filepath.FromSlash(dl.basePath + "/" + volumeName)
+}
+
+func (dl *DataLayer) variantPath(volumeName, variantName string) string {
+	return filepath.FromSlash(dl.basePath + "/" + volumeName + "/branches/" + variantName)
+}
+
+func (dl *DataLayer) commitPath(volumeName string, commitId CommitId) string {
+	return filepath.FromSlash(dl.basePath + "/" + volumeName + "/commits/" + string(commitId))
 }
 
 func (dl *DataLayer) CreateVolume(volumeName string) error {
-	volumePath := filepath.FromSlash(dl.BasePath + "/" + volumeName)
+	volumePath := dl.volumePath(volumeName)
 	return os.MkdirAll(volumePath, 0777)
 }
 
 func (dl *DataLayer) RemoveVolume(volumeName string) error {
-	volumePath := filepath.FromSlash(dl.BasePath + "/" + volumeName)
+	volumePath := filepath.FromSlash(dl.basePath + "/" + volumeName)
 	return os.RemoveAll(volumePath)
 }
 
 func (dl *DataLayer) CreateVariant(volumeName, variantName string) error {
 	// XXX Variants are meant to be tagged commits???
-	variantPath := filepath.FromSlash(dl.BasePath + "/" + volumeName + "/branches/" + variantName)
+	variantPath := filepath.FromSlash(dl.basePath + "/" + volumeName + "/branches/" + variantName)
 	return os.MkdirAll(variantPath, 0777)
 }
+
+func (dl *DataLayer) sanitizePath(path string) error {
+	// Calculate that dl.basePath is a strict prefix of filepath.Clean(path)
+	if !strings.HasPrefix(filepath.Clean(path), dl.basePath) {
+		return fmt.Errorf("%s is not a prefix of %s", filepath.Clean(path), dl.basePath)
+	}
+	return nil
+}
+
+func (dl *DataLayer) copyFiles(from, to string) error {
+	if err := dl.sanitizePath(from); err != nil {
+		return err
+	}
+	if err := dl.sanitizePath(to); err != nil {
+		return err
+	}
+
+	// check that ``from`` exists
+	sourceFile, err := os.Stat(from)
+	if err != nil {
+		return err
+	}
+	if !sourceFile.Mode().IsDir() {
+		return fmt.Errorf("%s is not a directory", from)
+	}
+	// check that ``to`` does not exist
+	_, err = os.Stat(to)
+	if err == nil {
+		return fmt.Errorf("%s already exists", to)
+	}
+	cmd := exec.Command("cp", "-a", from, to)
+	err = cmd.Run()
+	if err != nil {
+		return err
+	}
+	return err
+}
+
+func (dl *DataLayer) Snapshot(volumeName, variantName, commitMessage string) (CommitId, error) {
+	// TODO: also accept timestamp, user data
+	uuid1, err := uuid.NewV4()
+	if err != nil {
+		return CommitId(""), err
+	}
+	uuid2, err := uuid.NewV4()
+	if err != nil {
+		return CommitId(""), err
+	}
+	bigUUID := uuid1.String() + uuid2.String()
+	bigUUID = strings.Replace(bigUUID, "-", "", -1)
+	commitId := CommitId(bigUUID[:40])
+	variantPath := dl.variantPath(volumeName, variantName)
+	commitPath := dl.commitPath(volumeName, commitId)
+	if _, err := os.Stat(commitPath); err == nil {
+		return CommitId(""), fmt.Errorf("UUID collision. Please step out of the infinite improbability drive.")
```

Comment:
You are not passing any formatting to `fmt.Errorf` so `errors.New` is more appropriate. However more importantly, it seems like this is a spurious error that does not mean anything to the user. The UUID generation should just be retried until a unique UUID is generated.